### PR TITLE
V3.3.0.0-dev

### PR DESCRIPTION
**Pull request by nibbles83**
Here we go again.
CBlock updates, CChain refactoring & bugfixes galore.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -1329,30 +1436,62 @@ bool GetTransaction(const uint256 &hash, CTransaction &txOut, uint256 &hashBlock
...
+    // Read block
+    try {
+        filein >> block;
+    }
+    catch (std::exception &e) {
+        return error("%s() : deserialize or I/O error", __PRETTY_FUNCTION__);
+    }
+
+    // Check the header
+    if (block.IsProofOfWork() && !CheckProofOfWork(block.GetPoWHash(), block.nBits))
```
Comment (davidstoneham) [inline src/main.cpp:1483]: Should this also check POS work? Peercoin is a good reference since they are on btc core ~0.10

File: src/main.cpp [hunk 2]
```diff
@@ -1562,23 +1701,102 @@ int GetNumBlocksOfPeers()
...
+    // We define a condition which we should warn the user about as a fork of at least 7 blocks
+   // who's tip is within 72 blocks (+/- 12 hours if no one mines it) of ours
+   // We use 7 blocks rather arbitrarily as it represents just under 10% of sustained network
+   // hash rate operating on the fork.
+   // or a chain that is entirely longer than ours and invalid (note that this should be detected by both)
+   // We define it this way because it allows us to only store the highest fork tip (+ base) which meets
+   // the 7-block condition and from this always have the most-likely-to-cause-warning fork
+    if (pfork && (!pindexBestForkTip || (pindexBestForkTip && pindexNewForkTip->nHeight > pindexBestForkTip->nHeight)) &&
+        pindexNewForkTip->nChainTrust - pfork->nChainTrust > (CBigNum(pfork->GetBlockTrust()) * 7).getuint256() &&
+        chainActive.Height() - pindexNewForkTip->nHeight < 72)
```
Comment (davidstoneham) [inline src/main.cpp:1794]: I assume this should be increased since their block time is 10 minutes and ours is 1.5

File: src/main.cpp (hunk 1 above)
Comment (nibbles83) [reply src/main.cpp:1483]: They haven't refactored any of this so the code still resides in main.h however it the same as we have here.
https://github.com/peercoin/peercoin/blob/ebb4003ce8367501020181f7e734d52c4b1ab5ea/src/main.h#L1621

Comment (davidstoneham) [reply src/main.cpp:1483]: roger happy to keep this as is then

File: src/main.cpp (hunk 2 above)
Comment (nibbles83) [reply src/main.cpp:1794]: Just checked a number of other coins with varying block times, They all seem to retain these original fork warning condition values.

Comment (davidstoneham) [reply src/main.cpp:1794]: roger

**Pull request merged**